design/FY2019/Refactor and cache two psychrometric functions.md

```diff
@@ -0,0 +1,52 @@
+# Refactor and cache two psychrometric functions
+
+Xuan Luo, Tianzhen Hong
+
+Lawrence Berkeley National Laboratory
+
+July 25, 2019 
+
+## Justification for Feature Update
+
+The feature is to convert the funtion `PsyTsatFnHPb` from Fortran code to C++ code, and use caching for two psychrometric funtions - `PsyTsatFnHPb` & `PsyTsatFnPb`.
+
+The caching helps achieve run time performance gain but may trigger some small differences in the reported results. Table 1 summarizes the regression diffs of the two psychrometric functions. The summary is based on the regression test results with 623 EnergyPlus test files. The max diff in annual energy/facility report indicates the maximum relative difference in the summary tables, out of the total 623 tests. The math diffs indicate the differences in the reported variables and meters by hour or by time step. The tables diffs indicate the differences in the final report tables.
+
+**Table 1. Regression diffs of two psychrometric functions - annual**
+
+| ﻿Trade off flags       | Number of tests have large math diffs | Number of tests have large table diffs | Annual Site EUI   |                                | Annual Energy Meters Entire Facility - Electricity |                                | Annual Energy Meters Entire Facility - Gas |                                | Annual Energy Meters Entire Facility - Cooling |                                | Annual Energy Meters Entire Facility - Water |                                | Annual Energy Meters Entire Facility - Other |                                |
```

Comment:
This table _is_ helpful, but what would maximize the reviewability would be this table:

| File  | ESO Abs Diffs | ESO Rel Diffs | MTR Abs Diffs | MTR Rel Diffs | Extra Diffs | Notes |
|-------|---------------|---------------|---------------|---------------|-------------|-------|
| File1 | X             | X             | X             |               | Blah        | OK    |
| File2 | X             |               | X             |               | Blah        | OK    |
| File3 |               | X             | X             | X             | Blah        | OK    |

This is really what I was hoping for as it clearly itemizes which files have **real** diffs that are both absolute in nature as well as relative.  The table as it is right now has a lot of information, which is great, but it's tough to dig into.

Reply:
@Myoldmopar So the case was that all files, currently showing big diffs, show large abs and rel diff at the same time. I modified the table to only include files that, for one or more reported attributes, there are large values showing at large rel diff hours. And I summarized the number of hours in annual simulation this happens.

Some of them are actually NOT OK - but mostly due to
1. The threshold for large abs diff is always 0.01, 0.001, etc., no matter how large or small the attribute (e.g. mass flow rate) can reach, and for large rel diff is always 0.005.
2. The regression test has existing severe errors or recurring warnings.
3. Different Solver/Simulation Iteration Count 
4. Water heaters, small chiller evaporators, and ground heat exchangers have weird behaviors with some files.

I attached a detailed summary table in a spreadsheet but that is not necessarily intended to be included for release.

Reply:
> 1. The threshold for large abs diff is always 0.01, 0.001, etc., no matter how large or small the attribute (e.g. mass flow rate) can reach, and for large rel diff is always 0.005.

The mathdiff regressions use a special configuration that is dependent on the unit system for each value as well as the reporting frequency of the value.  

> 3. Different Solver/Simulation Iteration Count

Yep, this makes sense, I expect this type of impact.

> 4\. Water heaters, small chiller evaporators, and ground heat exchangers have weird behaviors with some files.

Wording like "have weird behaviors" makes me concerned about merging changes like this in.  If we don't know the impact of our changes except that they cause things to act weird, maybe they're not ready?